### PR TITLE
Introduce new intermediate representation and register allocator

### PR DESCRIPTION
**Pull request by vacantron**
As #79 mentioned, the stack-based virtual machine uses the available registers inefficiently in RISC-like architecture. The newly two-phase IR replaces the origin one which was coupled with the the register allocation and made the optimization harder.

The first phase IR uses the three-address code (3AC) to simplify the complex expression of C. The register allocator could use these information to generate the more efficient instructions, using the algorithm that supports the register-base virtual machine. It use the linear-scan algorithm in this implementation.

The second phase IR is going to locate the symbol which is used in branch/jump instruction. Once the amount of the other instructions is confirmed, the offset/address would be correct.

The new ARM backend uses register r0 ~ r7 for general purposes, and register r8 as the temporary storage for some instructions. Register r12 is as the global stack pointer for the convenience.

The new RISC-V backend uses register a0 ~ a7 for general purposes, and register t0 as the temporary storage for some instructions.

### CODE REVIEW

File: lib/c.c [hunk 1]
```diff
@@ -398,6 +463,7 @@ FILE *fopen(char *filename, char *mode)
 #endif
 
     abort();
+    return -1;
```
Comment (jserv) [inline lib/c.c:466]: Is this change necessary?

Comment (vacantron) [reply lib/c.c:466]: For the present, the function with non-void return type is expected to have a return statement. Maybe we could remove `abort()` and return `NULL`, and then check the `FILE` pointer be valid or not when using?

Comment (jserv) [reply lib/c.c:466]: Let's drop the call of `abort`.

File: src/arm-codegen.c [hunk 2]
```diff
@@ -1,532 +1,457 @@
 /* Translate IR to target machine code */
 
 #include "arm.c"
+#include "reg-alloc.c"
```
Comment (jserv) [inline src/arm-codegen.c:4]: Instead of the inclusion of `reg-alloc.c`, can you specify the header?

File: src/globals.c [hunk 3]
```diff
@@ -261,7 +307,7 @@ func_t *find_func(char func_name[])
 {
     int index = find_trie(FUNC_TRIES, func_name);
     if (index)
-        return &FUNCS[index - 1]; /* As the recorded index is a positive
+        return &FUNCS[index]; /* As the recorded index is a positive
```
Comment (jserv) [inline src/globals.c:310]: Refine the comment.

File: src/arm-codegen.c (hunk 2 above)
Comment (vacantron) [reply src/arm-codegen.c:4]: Is it better that moving `dump_ir()`and the inclusion of `reg-alloc.c` into `main.c`?

File: src/reg-alloc.c [hunk 4]
```diff
@@ -0,0 +1,843 @@
+/* Allocate register from IR. Using linear-scan algorithm */
```
Comment (jserv) [inline src/reg-alloc.c:1]: Mention the limitations as well.

File: lib/c.c [hunk 5]
```diff
@@ -346,23 +346,88 @@ void printf(char *str, ...)
...
+            if (str[si] >= '1' && str[si] <= '9') {
+                w = str[si] - '0';
+                si++;
+                if (str[si] >= '0' && str[si] <= '9') {
+                    w = w * 10;
+                    w += str[si] - '0';
+                    si++;
+                }
+            }
+            if (str[si] == 's') {
```
Comment (jserv) [inline lib/c.c:391]: Use switch-case for readable code.

Comment (vacantron) [reply lib/c.c:391]: May I change it with `printf()` together in another pull request?

File: src/defs.h [hunk 6]
```diff
@@ -104,32 +101,23 @@ typedef enum {
...
     char var_name[MAX_VAR_LEN];
     int is_ptr;
     int is_func;
     int array_size;
-    int offset;   /* offset from stack or frame */
+    int offset;   /* offset from stack or frame, index 0 is reserved */
     int init_val; /* for global initialization */
+
+    int is_global;
+    int eol; /* end-of-life */
```
Comment (jserv) [inline src/defs.h:119]: Shall it be named as "liveness"?

File: src/reg-alloc.c [hunk 7]
```diff
@@ -0,0 +1,844 @@
+/**
+ * Allocate register from IR. Using linear-scan algorithm, need at least 7
```
Comment (jserv) [inline src/reg-alloc.c:2]: Rewrite as following:
> Allocate registers from IR. The linear-scan algorithm now expects a minimum of 7 available registers (typical for RISC-style architectures).

File: lib/c.c (hunk 5 above)
Comment (jserv) [reply lib/c.c:391]: For newly-added code, it's better to make improvements sooner rather than requiring extra effort later on.

File: src/cfront.c [hunk 8]
```diff
@@ -835,68 +879,68 @@ void read_inner_var_decl(var_t *vd, int anon)
...
-         * create dummy parameters to put all on stack
-         */
-        for (; vn < MAX_PARAMS; vn++) {
-            strcpy(vds[vn].type_name, "int");
-            strcpy(vds[vn].var_name, "var_arg");
-            vds[vn].is_ptr = 1;
-        }
+        /* variadic function. Max 8 parameters are passed. */
+        fd->va_args = 1;
+        vn = 8;
```
Comment (jserv) [inline src/cfront.c:907]: Use `MAX_PARAMS` when possible.

**Comment by jserv** [comment]
Is it feasible to close #22 with register allocation?

**Comment by vacantron** [comment]
> Is it feasible to close #22 with register allocation?

Yes. I update the limitation of the operand stack size in [b41e76c](https://github.com/sysprog21/shecc/pull/80/commits/b41e76cf4809f410d8a51d165a045b148f7a0a10) . It can pass the test now.

**Comment by jserv** [comment]
@vacantron, Use `git rebase -i` to rework these git commits, including updating the commit messages, which may include statements like `Close #22`. I am preparing to merge this pull request.

**Comment by jserv** [comment]
Thank @vacantron for contributing!
